qiskit/utils/deprecate_pulse.py

```diff
@@ -0,0 +1,90 @@
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+Deprecation functions for Qiskit Pulse. To be removed in Qiskit 2.0.
+"""
+
+import warnings
+import functools
+
+from qiskit.utils.deprecation import deprecate_func, deprecate_arg
+
+
+def deprecate_pulse_func(func):
+    """Deprecation message for functions and classes"""
+    return deprecate_func(
+        since="1.3",
+        package_name="Qiskit",
+        removal_timeline="in Qiskit 2.0",
+        additional_msg="The entire Qiskit Pulse package is being deprecated "
+        "and will be moved to the Qiskit Dynamics repository: "
+        "https://github.com/qiskit-community/qiskit-dynamics",
+    )(func)
+
+
+def deprecate_pulse_dependency(*args, **kwargs):
+    """Deprecation message for functions and classes which use or depend on Pulse"""
+    decorator = deprecate_func(
+        since="1.3",
+        package_name="Qiskit",
+        removal_timeline="in Qiskit 2.0",
+        additional_msg="The entire Qiskit Pulse package is being deprecated "
+        "and this is a dependency on the package.",
+        **kwargs,
+    )
+
+    # Taken when `deprecate_pulse_dependency` is used with no arguments and with empty parentheses,
+    # in which case the decorated function is passed
+    if args:
+        return decorator(args[0])
+    return decorator
+
+
+def deprecate_pulse_arg(arg_name, **kwargs):
+    """Deprecation message for arguments related to Pulse"""
+    return deprecate_arg(
+        name=arg_name,
+        since="1.3",
+        package_name="Qiskit",
+        removal_timeline="in Qiskit 2.0",
+        additional_msg="The entire Qiskit Pulse package is being deprecated "
+        "and this argument uses a dependency on the package.",
+        **kwargs,
+    )
+
+
+def ignore_pulse_deprecation_warnings(func):
+    """Ignore deprecation warnings emitted from the pulse package"""
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        with warnings.catch_warnings():
+            warnings.filterwarnings(
+                "ignore", category=DeprecationWarning, message="The (.*) ``qiskit.pulse"
+            )
+            return func(*args, **kwargs)
+
+    return wrapper
+
+
+def decorate_test_methods(decorator):
+    """Put a given decorator on all the decorated class methods whose name starts with `test_`"""
+
+    def cls_wrapper(cls):
+        for attr in dir(cls):
+            if attr.startswith("test_") and callable(object.__getattribute__(cls, attr)):
+                setattr(cls, attr, decorator(object.__getattribute__(cls, attr)))
+
+        return cls
+
+    return cls_wrapper
```

Comment:
In my experience, we usually tackle these large deprecations adding filters to the qiskit test base class rather than defining a custom decorator for this: https://github.com/Qiskit/qiskit/blob/fbfe7388f3efdaaeb70dc96aa35f649c1897c5d4/test/utils/base.py#L171-L177. It is less granular than your approach (in the sense that we don't control which tests have raised these filtered warnings) but I think it makes the removal process smoother. This is not standardized and it probably will depend a bit on who you ask.

Reply:
There is a tradeoff here... It seems that the preferred way in Qiskit to handle deprecation warnings in testing is by using `unittest.assertWarns` and that `QiskitTestCase`-level filtering is meant to be used mainly for messages emitted by 3rd party packages or otherwise in very special circumstances like deprecating a key core element which makes the use of `assertWarns` really problematic.
In this case, I wanted to avoid putting `assertWarns` everywhere in `test/python/pulse` but still use it on deprecated functions outside of Pulse like `QuantumCircuit.calibrations`. So I took a hybrid approach where I used the `@decorate_test_methods(ignore_pulse_deprecation_warnings)` only in `test/python/pulse` and `test/python/visualization_v2` which both are being deprecated as a whole, as well as on the calibration builder passes in `test/python/transpiler` which are being deprecated as well. I like this balance but I'm open to other views or ideas about how to achieve this in a different way. 
BTW, FWIW, these decorators are useful as markers for things to be removed in 2.0.